extend-payload-to-esp:  extends any src_input_dir to esp

say-paul opened the pull request:
should fix: #651 
 fixes : #766 
 To support raspberry pi 4, which works well for rpm-ostree, but in bootc.
 This PR add command `extend-payload-to-esp`  which will enable bootup install to move the files to esp.
usage:
```
RUN dnf install -y uboot-images-armv8

# rpm -ql uboot-images-armv8 | grep rpi
# /usr/share/uboot/rpi_arm64
# /usr/share/uboot/rpi_arm64/u-boot.bin

RUN bootupctl backend extend-payload-to-esp /usr/share/uboot/rpi_arm64
```
 
what extend-payload-to-esp does:
- stages the files to /usr/lib/efi/firmware/{name}/{version}/EFI/
- stores rpm details and timestamp in /usr/lib/efi/firmware/{name}/{version}/EFI.json
```
bash-5.2# tree /usr/lib/efi/firmware/uboot/
/usr/lib/efi/firmware/uboot/
`-- 1:2025.04-1
    |-- EFI
    |   `-- u-boot.bin
    `-- EFI.json
```
- `bootupd install` and `bootupd update` is updated to read from the above path and install to esp.
- use the metadata(EFI.json) to update the bootupd-state

rpi4 -usecase: https://github.com/fedora-iot/iot-distro/issues/82#issuecomment-3051419262

cgwalters reviewed (dismissed):
Thanks for working on this!

say-paul commented on src/efi.rs:
```diff
@@ -162,6 +162,39 @@ impl Efi {
+
+    /// Copy firmware files from /boot/efi to target directory
+    fn copy_firmware_blobs(&self, src_root: &openat::Dir, dest_root: &str) -> Result<()> {
+        let src_efi = "boot/efi";
```
Ran: #766 with he current implmentation with an additional step like of creating the files in ` /boot`:
```
# Build from the current git into a c9s-bootc container image.
# Use e.g. --build-arg=base=quay.io/fedora/fedora-bootc:41 to target
# Fedora or another base image instead.
#
ARG base=quay.io/centos-bootc/centos-bootc:stream9

FROM $base as build
# This installs our package dependencies, and we want to cache it independently of the rest.
# Basically we don't want changing a .rs file to blow out the cache of packages.
RUN <<EORUN
set -xeuo pipefail
dnf -y install cargo git openssl-devel
EORUN

RUN mkdir -p /boot/firmware/efi && \
    touch /boot/firmware/efi/dummy.{dtb,dat,efi,elf,bin}
# Now copy the source
COPY . /build
WORKDIR /build
# See https://www.reddit.com/r/rust/comments/126xeyx/exploring_the_problem_of_faster_cargo_docker/
# We aren't using the full recommendations there, just the simple bits.
RUN --mount=type=cache,target=/build/target --mount=type=cache,target=/var/roothome \
    make && make install-all DESTDIR=/out
  
# FROM $base
# # Clean out the default to ensure we're using our updated content
# RUN rpm -e bootupd
# COPY --from=build /out/ /
# # Sanity check this too
# RUN bootc container lint --fatal-warnings
  
    
``` 
From #755 
> Content shipped in an RPM with files in /boot that are present in the initial base image build where bootupctl backend generate-update-payload is invoked

The build container has the dummy files:
```
bash-5.2# ls /boot/firmware/efi/          
dummy.bin  dummy.dat  dummy.dtb  dummy.efi  dummy.elf
```

but the final build bu using the copy commands  doesn't have the files,  :confused:

say-paul replied:
I think this function can be modified into a generic one to copy files, but I am not sure how to trigger it from the container build process

cgwalters replied:
> but the final image that is built using the COPY --from=build /out/ / doesn't have the files, 😕

The `build` image is only intended to hold the result of building the bootupd binary itself. If you want to modify the final image for other reasons (e.g. injecting kernel arguments, adding debug tools or other configuration, and as in this case inject arbitrary other content) then you should put it on the last stage.

Or better even - don't modify the dockerfile, make a new one and have it be `FROM localhost/bootupd`.

say-paul commented:
It should now stage the updates for any user input path,
I think we have to additionally chain this in  efi::{install,run_update}
Its very similar to `generate_update_metadata`  but with an user_input.

travier commented:
A lot of questions are unanswered here:
- Is this file versioned?
- Is it always installed under the same name?
- Is it tied to a kernel version?
- What happens when we rollback?

Overall, I would suggest we work toward https://github.com/coreos/bootupd/issues/926 which should get us to https://github.com/coreos/bootupd/issues/766.

say-paul commented:
@travier please let me know if I am missing anything.

cgwalters commented:
Can you add an example canonical usage of this in the initial PR comment or so at least? A test would be even more ideal.

Basically to demonstrate how this is used.

I think again just to summarize two important goals:
- no changes to `bootc install` are required
- `bootupctl upgrade` Just Works to also upgrade

say-paul commented:
I can add test to it, but I want to validate my approach.
I ran the following to test it manually:

I usedc Containerfile:
```
ARG base=quay.io/fedora/fedora-bootc:41

FROM $base as build
# This installs our package dependencies, and we want to cache it independently of the rest.
# Basically we don't want changing a .rs file to blow out the cache of packages.
RUN <<EORUN
set -xeuo pipefail
dnf -y install cargo git openssl-devel
EORUN
# Now copy the source
COPY . /build
WORKDIR /build
# See https://www.reddit.com/r/rust/comments/126xeyx/exploring_the_problem_of_faster_cargo_docker/
# We aren't using the full recommendations there, just the simple bits.
RUN --mount=type=cache,target=/build/target --mount=type=cache,target=/var/roothome \
    make && make install-all DESTDIR=/out

FROM $base
# Clean out the default to ensure we're using our updated content
RUN rpm -e bootupd
COPY --from=build /out/ /
RUN dnf -y install uboot-images-armv8
RUN bootupctl backend extend-payload-to-esp  /usr/share/uboot/rpi_arm64
```
Build on rpi4:
`[root@192 bootupd]# podman build -t bootc_payload_extend --build-arg=base=quay.io/fedora/fedora-bootc:42 .`

Validation:
```
[root@192 bootupd]# podman run -it localhost/bootc_payload_extend:latest /bin/bash
bash-5.2# tree /usr/lib/efi/firmware/
/usr/lib/efi/firmware/
`-- uboot
    `-- 1:2025.04-2
        `-- EFI
            `-- boot
                `-- efi
                    `-- u-boot.bin
```

the final destination is in accordance to #926 , #938
I am unaware of how the things proceeds next,  but i think generate-update-metadata() from #938 can pick it up and process it to for installation.

say-paul commented:
Extended install to copy from the extend-payload.
 Build the image with bib, I could install uboot to esp.

```
λ sudo fdisk -l /dev/nbd0
Disk /dev/nbd0: 10 GiB, 10737418240 bytes, 20971520 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: D209C89E-EA5E-4FBD-B161-B461CCE297E0

Device        Start      End  Sectors  Size Type
/dev/nbd0p1    2048  1028095  1026048  501M EFI System
/dev/nbd0p2 1028096  3125247  2097152    1G Linux filesystem
/dev/nbd0p3 3125248 20971486 17846239  8.5G Linux filesystem
λ mkdir ~/qcow2-mount
sudo mount /dev/mapper/nbd0p1 ~/qcow2-mount
[~/git_repos/bootupd]$(testing-frimware-copy)
λ ls -l ~/qcow2-mount
total 688
drwxr-xr-x. 4 root root   8192 Jul  5 05:56 EFI
-rwxr-xr-x. 1 root root 690712 Jul  5 05:56 u-boot.bin
[~/git_repos/bootupd]$(testing-frimware-copy)
λ ls -al ~/qcow2-mount
total 704
drwxr-xr-x. 3 root    root     16384 Jan  1  1970 .
drwx--x---+ 1 saypaul saypaul   2038 Jul  5 06:17 ..
drwxr-xr-x. 4 root    root      8192 Jul  5 05:56 EFI
-rwxr-xr-x. 1 root    root    690712 Jul  5 05:56 u-boot.bin
[~/git_repos/bootupd]$(testing-frimware-copy)
```

say-paul commented:
bootupd-state.json: (adds 'firmware' to EFI) - probably should not have  EFI.json in fireware::filetree, which should be a minor fix.

```
λ cat ~/qcow2-mount-p2/bootupd-state.json | jq
{
  "installed": {
    "EFI": {
      "meta": {
        "timestamp": "2025-03-27T10:27:59Z",
        "version": "grub2-efi-aa64-1:2.12-28.fc42.aarch64,shim-aa64-15.8-3.aarch64"
      },
      "filetree": {
        "children": {
          "BOOT/BOOTAA64.EFI": {
            "size": 971654,
            "sha512": "sha512:8891ba415802b97828b7939b81c0235610f3648fd257df8e4b754e1fa4899da9e202fd1be95bbcb095816525b601a6ecfa3ec94be08de5f5fa1911cffc683d13"
          },
          "BOOT/fbaa64.efi": {
            "size": 91096,
            "sha512": "sha512:a1289697eb9686c171446fd7ae2b1fd7358b7aa8de229f6d22993852f4f5be41f0b783f50b33505167014e85b07deadd6fed2533c68eb4577743bf6c5df61ff3"
          },
          "fedora/BOOTAA64.CSV": {
            "size": 112,
            "sha512": "sha512:4ceedbfec87e417308a71313d22b6b81aa3b00c70f55d35c18bd97e88d45c0429032cb822b31340a8c105e47bb45d4c44871d6de29bb397d324b735752c42b7c"
          },
          "fedora/grubaa64.efi": {
            "size": 4210160,
            "sha512": "sha512:2e84d665dbac56f4b0fa369023841c8e8c16bd0d68f0f3bcbfdfeae7236f5396e03d1852238d85b3a2e0d3c4f64d443bf66d5f79e1fa22e304440a9a72fb3e7c"
          },
          "fedora/mmaa64.efi": {
            "size": 886296,
            "sha512": "sha512:fda90b956737d74b6d66b491b66573bd47f857da1a2ef8e619167173b1b870991b0a904961714462794dcaebfbd0d026c0bd8afdfcdb21932f0a7c43163855d4"
          },
          "fedora/shim.efi": {
            "size": 971654,
            "sha512": "sha512:8891ba415802b97828b7939b81c0235610f3648fd257df8e4b754e1fa4899da9e202fd1be95bbcb095816525b601a6ecfa3ec94be08de5f5fa1911cffc683d13"
          },
          "fedora/shimaa64.efi": {
            "size": 971654,
            "sha512": "sha512:8891ba415802b97828b7939b81c0235610f3648fd257df8e4b754e1fa4899da9e202fd1be95bbcb095816525b601a6ecfa3ec94be08de5f5fa1911cffc683d13"
          }
        }
      },
      "adopted-from": null,
      "firmware": {
        "uboot": {
          "meta": {
            "timestamp": "2025-04-20T21:03:54Z",
            "version": "uboot-images-armv8-1:2025.04-2.fc42.noarch"
          },
          "filetree": {
            "children": {
              "EFI.json": {
                "size": 91,
                "sha512": "sha512:09f2147475392cea035a408e269c13dcf8d7ef2d3c7a9fd34328131de179c14c2dd27510f3c22807d571bfa7ad151a1b0787ad0b5a4113be3dfc98a4d8951c3f"
              },
              "u-boot.bin": {
                "size": 690712,
                "sha512": "sha512:10fe3903b2f94a3e3c1e7e95009388c2628c50a84ebe98d8c43f370856a0250feaa44044df92bea293a1612703e253e8d9d4a60192d30d355a2958133986fafe"
              }
            }
          },
          "adopted-from": null
        }
      }
    }
  },
  "pending": null,
  "static-configs": {
    "timestamp": "2025-07-04T23:50:49Z",
    "version": "0.2.28"
  }
}
```

say-paul commented:
@cgwalters I have updated the  comments [here](https://github.com/coreos/bootupd/pull/935#issue-3093610029) with the details + example.

kfox1111 commented:
Curious about a mode dealing with a rpi config.txt. There a way to have it land one and then not update it if its there already or alternately only if unchanged?

HuijingHei commented:
Since Colin and Timothee are busy with other things, I can help to see this. Let me know if I misunderstood it.

So does this mean that you only want to extend single file `/usr/share/uboot/rpi_arm64/u-boot.bin`,  not files under `/boot/efi` like `bcm2711-firmware`, is this right?

say-paul commented:
> So does this mean that you only want to extend single file /usr/share/uboot/rpi_arm64/u-boot.bin, not files under /boot/efi like bcm2711-firmware, is this right?

@HuijingHei On high level i want to copy content of a dir to /boot/efi that includes both containers: /boot/efi  and /usr/share/uboot/rpi_arm64/ to esp of the image.

this is the complete usecase : https://github.com/fedora-iot/iot-distro/issues/82#issuecomment-3051419262

HuijingHei commented:
> > So does this mean that you only want to extend single file /usr/share/uboot/rpi_arm64/u-boot.bin, not files under /boot/efi like bcm2711-firmware, is this right?
> 
> @HuijingHei On high level i want to copy content of a dir to /boot/efi that includes both containers: /boot/efi and /usr/share/uboot/rpi_arm64/ to esp of the image.
> 
> this is the complete usecase : [fedora-iot/iot-distro#82 (comment)](https://github.com/fedora-iot/iot-distro/issues/82#issuecomment-3051419262)

Agree with Colin and Timothee that there are partly duplication with https://github.com/coreos/bootupd/issues/926, e.g. `stages the input files to /usr/lib/efi/firmware/{name}/{version}/EFI/`, maybe we should make them independent functions and let both can reuse it. Would you like to change this (or I will do the change)? And it would be better to add test for [comment](https://github.com/fedora-iot/iot-distro/issues/82#issuecomment-3051419262) in the PR. WDYT?

say-paul committed:
feat(extend_payload_to_esp):stage payload to /usr for installation
extend_payload_to_esp <path> will move the all contents of the <path> to /usr

This features gives user ability to stage payloads so that install and update
can  install the contents to esp.It will move all the files and sub dirs
under the path provide by user to /usr/lib/efi/firmware/<name>/<version>/EFI/
Incase of an exiting path extend-payload will only keep the latest
version of the payload.
example usage: `bootupctl backend extend_payload_to_esp /usr/lib/mydata`
will move the content of the dir to
`/usr/lib/efi/firmware/<pkg-name>/<version>/EFI/`
stores metadata at `/usr/lib/efi/firmware/<pkg-name>/<version>/EFI.json`.
Added test_extend_payload_to_esp unit test.

say-paul committed:
update(efi::install)install from extend-payload-path
Added additional flow in efi::install to read from the path
created by extend_payload_to_esp command and install to /boot/efi

say-paul committed:
update(efi::update)update extended payload path
efi::update is updated to read from path created by
extend_payload_to_esp and update the firmware.

bam80 commented:
Anything left?

say-paul committed:
ci: added tests for install and update for extended payload

say-paul committed:
verbose

cgwalters commented on .github/workflows/ci.yml:
```diff
@@ -139,3 +139,196 @@ jobs:
+        run: |
+          set -xeuo pipefail
+          sudo truncate -s 5G myimage-extend.raw
+          sudo podman run --rm --privileged -v .:/target --pid=host --security-opt label=disable \
```
Actually thinking about this a bit more, why can't this test work how we'd expect things to work in production?

Basically if we change this to be a `podman build` that installs the files instead that should all Just Work, right?

cgwalters commented on src/efi.rs:
```diff
@@ -860,4 +1103,170 @@ Boot0003* test";
+                panic!("extend_payload returned None - expected success");
+            }
+            Err(e) => {
+                panic!("extend_payload failed when it should have succeeded: {}", e);
```
It'd avoid all this nesting to just use `.unwrap()` or `.expect()` on the result which seems preferable to me.
And we may as well do the same for the `Option`

cgwalters commented on src/efi.rs:
```diff
@@ -860,4 +1103,170 @@ Boot0003* test";
+
+        let original_path = std::env::var("PATH").unwrap_or_default();
+        let new_path = format!("{}:{}", mock_rpm_dir.path().display(), original_path);
+        std::env::set_var("PATH", &new_path);
```
This is fully `unsafe` for good reason in Rust 2024: https://doc.rust-lang.org/edition-guide/rust-2024/newly-unsafe-functions.html#stdenvset_var-remove_var

It's more painful but basically if we wanted to mock RPM we'd have to explicitly pass it as a parameter.

cgwalters commented on src/efi.rs:
```diff
@@ -506,6 +656,99 @@ impl Component for Efi {
+                anyhow!("RPM query returned an empty or malformed version string")
+            })?;
+
+        let parts: Vec<&str> = version_string_part.split('-').collect();
```
Hmm @HuijingHei is reworking the version parsing in https://github.com/coreos/bootupd/pull/977 etc.

cgwalters commented on src/efi.rs:
```diff
@@ -506,6 +656,99 @@ impl Component for Efi {
+                format!(
+                    "{}-{}",
+                    parts[parts.len() - 2],
+                    parts[parts.len() - 1]
```
Try using e.g.

```
match parts.as_slice() {
  [name, version, release] => { ... }
  _ => anyhow::bail!()
```

cgwalters commented on src/efi.rs:
```diff
@@ -409,14 +466,106 @@ impl Component for Efi {
             filetree::apply_diff(&updated, &destdir, &diff, None)
                 .context("applying filesystem changes")?;
+            // update firmware
+            let firmware_base_dir_path = Path::new("usr/lib/efi/firmware");
```
Where does this path come from? Is it something you made up here?

cgwalters commented on src/efi.rs:
```diff
@@ -377,15 +379,70 @@ impl Component for Efi {
+        let mut found_firmware = BTreeMap::new();
+        // Scan and install supplemental firmware
+        let firmware_base_dir_path = Path::new("usr/lib/efi/firmware");
+        if src_root.exists(firmware_base_dir_path)? {
```
The code past this is getting really deeply nested, can we factor out the inner part here into a helper function?

cgwalters reviewed (commented):
Sorry for the delay in review and thanks for working on this.

I have some initial comments.

alexlarsson commented:
> We had a chat a while back and this should now be possible with what we have in #938.

@travier Can #938 be used to copy firmware to the root of the esp? It looks like it can only copy into the EFI subdirectory?

martinezjavier commented:
@say-paul I've proposed https://github.com/coreos/bootupd/pull/1073 that I think complements your `extend-payload-to-esp` sub-command
